### DISCUSSION TITLE
Problem with nested foreach manually throwing an error

### DISCUSSION THREAD

**Discussion by tiec7** [cypress-io/cypress#17594] [Questions and Help]
Hi, I'm getting mad with an issue I'm facing using Cypress; to avoid a wall of text I put my issue on stackoverflow here:

https://stackoverflow.com/questions/68648400/cypress-manually-throwing-error-in-a-nested-loop-that-does-multiple-checks

Does anyone has ANY idea or tip or anything about this?

Thanks in advance!

PS: obviously if you need to see more about code there will be no problem: :)

**Comment by portyu9** [comment]
Hi @tiec7 

This isn't really a nested `forEach` problem. You're effectively implementing **soft assertions** manually: catching each Chai assertion failure, continuing the test, collecting the errors, then failing once at the end.

I would avoid calling Cypress commands such as `cy.log()` inside those `catch` blocks. Keep the validation function synchronous, collect only the error messages, and fail once after all Cypress commands have completed.

For example:

```js
const errors = []

cy.wrap(ids)
  .each((id) => {
    return cy.get(id).then(($el) => {
      const result = checkAllWebElementsOfCityUrl($el, elem, id)

      if (result?.length) {
        errors.push(...result)
      }
    })
  })
  .then(() => {
    expect(
      errors,
      `Errors found:\n${errors.join('\n')}`
    ).to.be.empty
  })
```

Also, in your updated example use:

```js
cy.wrap(ids).each(...)
```

not:

```js
cy.get(ids).each(...)
```

because `cy.get()` is for querying a selector/alias, not iterating a JavaScript array.

The final `throw new Error(...)` itself is not accumulating work into later tests—it simply causes Cypress to enter its normal failure path. When running with `cypress run`, that also includes automatic failure handling such as screenshots.

So if later tests become progressively slower **only after failures**, I would treat that as a separate Cypress performance/version issue rather than a loop-control issue. As a diagnostic, try disabling `screenshotOnRunFailure` and upgrading Cypress. But structurally, the important fix is: **don't mix Cypress commands into the caught-assertion loop; collect synchronously and perform one final assertion.**

**Reply by portyu9** [reply]
@tiec7 Please mark this answer as accepted if it helped, thank you.
